### DISCUSSION TITLE
Possibility of switching core/table from static to dynamic block?

### DISCUSSION THREAD

**Discussion by Sidsector9** [WordPress/gutenberg#55089] [Developer Experience]
## What problem does this address?

Hello all, I wanted to discuss the possibility of converting/migrating the core Table block from static to dynamic type, and I am not sure whether this is possible, or at least discouraged.

There is a list of feature requests & enhancements for the Table block here - https://github.com/WordPress/gutenberg/issues/32400

Recently, I picked up the feature request to merge/un-merge cells here - https://github.com/WordPress/gutenberg/issues/15821
and made decent progress:

https://github.com/WordPress/gutenberg/assets/17757960/8d68e424-edf1-4066-b01e-b07b78454a04

This works pretty well until you save and reload. If you try to merge cells after reloading, it will break.
The reason is, when the HTML is saved to the DB, it excluded the cells that were part of the merged cell. Due to this, when the editor reconstructs the table using the `save` function, the original index values are lost.

I found it to be extremely complicated to fix it as the context of the table before and after saving is lost. And reconstruction passes values to the attributes after parsing the HTML of the saved data.

If we were to use a dynamic table, we could store this information as Block attributes and then use it to build the table server-side.

## What is your proposed solution?

I have more of a concept at the moment rather than the actual solution. If this is feasible then I'll be more than willing to contribute to the discussion.

**Comment by t-hamano** [comment]
Hi, @Sidsector9,

The core table block supports [colSpan and rowSpan attributes](https://github.com/WordPress/gutenberg/blob/b8b89cecd6b6e274d70867d9d0dd7acd005be3e5/packages/block-library/src/table/block.json#L51-L60). Therefore, if the logic that merge/un-merge cells properly controls those attributes, the block should not be broken.

Additionally, I think you should only create dynamic blocks where the content changes due to external factors, such as a site title block, or when the content changes when the block is rendered, such as a recent posts block.

I think another reason not to make blocks dynamic is to avoid reducing performance during rendering.

By the way, I have created [a custom table block](https://github.com/t-hamano/flexible-table-block) in the past to try the cell merging/unmerging function. This plugin is also published in the plugin directory.

**Reply by spencerfinnell** [reply]
@t-hamano your plugin has been very helpful 😄

**Reply by Sidsector9** [reply]
@t-hamano I explored your code-base and it's impressive. I did not think of an approach using a Virtual Table and handling it that way. I think I understand why it was so difficult for me to work with my implementation. Thanks for sharing!

**Comment by talldan** [comment]
> The reason is, when the HTML is saved to the DB, it excluded the cells that were part of the merged cell. Due to this, when the editor reconstructs the table using the save function, the original index values are lost.

The demo in the video looks really cool!

The problem you mention is a really interesting one. It'd be really useful to have a PR to reference showing the issue so we can discuss potential solutions, as it's hard to visualize what's happening without being able to debug it.

**Reply by Sidsector9** [reply]
@talldan thank you :) 

Up until I saw @t-hamano's implementation I thought it was impossible to merge/unmerge using static a table. I will raise the PR by tomorrow and share the implementation I have, in the mean time I will also try to figure out how to best integrate Aki's approach by using a virtual table.

**Reply by talldan** [reply]
Thanks!
